### PR TITLE
Tix/delta perf hook

### PR DESCRIPTION
**Pull request by Tixxx**

### CODE REVIEW

**Comment by kit1980** [comment]
Seems like normal DistributeOptimizer is broken:

```
Traceback (most recent call last):
  File "pytorch/train.py", line 249, in <module>
    train_htk()
  File "pytorch/train.py", line 95, in train_htk
    backward_passes_per_step=args.backward_passes_per_step,
  File "/home/sergii/anaconda3/envs/parasail_new/lib/python3.7/site-packages/horovod-0.18.1-py3.7-linux-x86_64.egg/horovod/torch/__init__.py", line 330, in DistributedOptimizer
    return cls(optimizer.param_groups, named_parameters, compression, backward_passes_per_step, op)
TypeError: __init__() takes from 4 to 5 positional arguments but 6 were given
Traceback (most recent call last):
  File "pytorch/train.py", line 249, in <module>
    train_htk()
  File "pytorch/train.py", line 95, in train_htk
    backward_passes_per_step=args.backward_passes_per_step,
  File "/home/sergii/anaconda3/envs/parasail_new/lib/python3.7/site-packages/horovod-0.18.1-py3.7-linux-x86_64.egg/horovod/torch/__init__.py", line 330, in DistributedOptimizer
    return cls(optimizer.param_groups, named_parameters, compression, backward_passes_per_step, op)
TypeError: __init__() takes from 4 to 5 positional arguments but 6 were given
```

**Comment by kit1980** [comment]
Seems like normal DistributeOptimizer is broken:

```
Traceback (most recent call last):
  File "pytorch/train.py", line 249, in <module>
    train_htk()
  File "pytorch/train.py", line 95, in train_htk
    backward_passes_per_step=args.backward_passes_per_step,
  File "/home/sergii/anaconda3/envs/parasail_new/lib/python3.7/site-packages/horovod-0.18.1-py3.7-linux-x86_64.egg/horovod/torch/__init__.py", line 330, in DistributedOptimizer
    return cls(optimizer.param_groups, named_parameters, compression, backward_passes_per_step, op)
TypeError: __init__() takes from 4 to 5 positional arguments but 6 were given
Traceback (most recent call last):
  File "pytorch/train.py", line 249, in <module>
    train_htk()
  File "pytorch/train.py", line 95, in train_htk
    backward_passes_per_step=args.backward_passes_per_step,
  File "/home/sergii/anaconda3/envs/parasail_new/lib/python3.7/site-packages/horovod-0.18.1-py3.7-linux-x86_64.egg/horovod/torch/__init__.py", line 330, in DistributedOptimizer
    return cls(optimizer.param_groups, named_parameters, compression, backward_passes_per_step, op)
TypeError: __init__() takes from 4 to 5 positional arguments but 6 were given
```

File: horovod/torch/__init__.py [hunk 1]
```diff
@@ -209,69 +208,181 @@ class _DistributedAdasumOptimizer(torch.optim.Optimizer):
...
+        # compute delta = curr - start
+        p.data.sub_(start)
+        
+        # allreduce as before
+        tensor_compressed, ctx = self._compression.compress(p)
+        handle = allreduce_async_(tensor_compressed.data, name=name, op=Adasum)
+
+        # reset stashed parameters
+        for stashed, group in zip(stashed_params, self.param_groups):
+            group['params'] = stashed
```
Comment (jaliyae) [inline horovod/torch/__init__.py]: >         [](start = 37, length = 8)

nit: extra space

File: horovod/torch/__init__.py [hunk 2]
```diff
@@ -209,69 +208,181 @@ class _DistributedAdasumOptimizer(torch.optim.Optimizer):
...
+        missing_p = self._requires_update - set(self._handles.keys())
+        for p in missing_p:
+            handle, ctx = self._allreduce_grad_async(p)
+            self._handles[p] = (handle, ctx)
+
+        for p, (handle, ctx) in self._handles.items():
+            if not handle:
+                raise AssertionError("optimizer.zero_grad() was called after loss.backward() "
+                                    "but before optimizer.step() or optimizer.synchronize(). "
+                                    "This is prohibited as it can cause a race condition.")
```
Comment (jaliyae) [inline horovod/torch/__init__.py]: > race condition." [](start = 73, length = 17)

Do we even get to this or a hang?, what happen in default horovod?

File: horovod/torch/__init__.py [hunk 3]
```diff
@@ -209,69 +208,181 @@ class _DistributedAdasumOptimizer(torch.optim.Optimizer):
...
+            delta = synchronize(handle)
+            delta = self._compression.decompress(delta, ctx)
```
Comment (jaliyae) [inline horovod/torch/__init__.py]: >             [](start = 60, length = 12)

nit:extra space

**Pull request merged**